MINOR: Refactor KRaft image generation

cmccabe opened the pull request:
Refactor how we generate KRaft metadata images from MetadataImage.java. This is a step in the direction of properly supporting lossy downgrade, although more steps will be needed.

This PR adds a new ImageWriter interface which replaces the generic Consumer interface which accepted lists of records. It is better to do batching in the ImageWriter than to try to deal with that complexity in the MetadataImage#write functions, especially since batching is not semantically meaningful in KRaft snapshots. The new ImageWriter interface also supports freeze and close, which more closely matches the semantics of the underlying Raft classes.

The PR also adds an ImageWriterOptions class which we can use to pass parameters to control how the new image is written. Right now, the parameters that we are interested in are the target metadata version (which may be more or less than the original image's version) and a handler function which is invoked whenever metadata is lost due to the target version.

Convert over the MetadataImage#write function (and associated functions) to use the new ImageWriter and ImageWriterOptions. In particular, we now have a way to handle metadata losses by invoking ImageWriterOptions#handleLoss.

Get rid of the use of SOME_RECORD_TYPE.highestSupportedVersion() in several places. In general, we do not want to "silently" change the version of a record that we output, just because a new version was added.  We should be explicit about what record version numbers we are outputting.

Move the highest offset, highest epoch, and highest timestamp of an image into ImageProvenance.java.
Remove them from MetadataDelta since many ways of constructing a delta don't require incrementally
tracking these (such as loading a snapshot)

Implement ProducerIdsDelta#toString, to make debug logs look better.

Move MockRandom.java from metadata/src/test into server-common/src/test since it is generally useful.

jsancio commented on core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala:
```diff
@@ -390,15 +389,18 @@ class BrokerMetadataListener(
     override def run(): Unit = {
-      _image.write(this)
-      future.complete(records)
+      val writer = new InMemoryImageWriter()
+      val options = new ImageWriterOptions.Builder().
+        setMetadataVersion(_image.features().metadataVersion()).
+        build()
+      try {
+        _image.write(writer, options)
+      } finally {
+        writer.close()
+      }
```
How about using Java's "try with resource"?

cmccabe replied:
try-with-resource does not exist in Scala

jsancio commented on metadata/src/main/java/org/apache/kafka/image/MetadataImage.java:
```diff
@@ -120,17 +116,17 @@ public AclsImage acls() {
+        clientQuotas.write(writer, options);
+        producerIds.write(writer, options);
+        acls.write(writer, options);
+        writer.freeze();
```
It feels awkward that this layer or code calls `freeze` but `close` is called by the layer above. I am leaning toward moving this call to `freeze` to the caller(s).

cmccabe replied:
I have revised this code so that there is just a close function which takes a boolean. We call it here.

cmccabe replied:
I have revised this code so that there is just a close function which takes a boolean. We call it here.

jsancio commented on metadata/src/main/java/org/apache/kafka/image/AclsImage.java:
```diff
@@ -53,13 +49,15 @@ public Map<Uuid, StandardAcl> acls() {
-    public void write(Consumer<List<ApiMessageAndVersion>> out) {
-        List<ApiMessageAndVersion> batch = new ArrayList<>();
+    public void write(ImageWriter writer, ImageWriterOptions options) {
+        // Technically, AccessControlEntryRecord appeared in 3.2-IV0, so we should not write it if
+        // the output version is less than that. However, there is a problem: pre-production KRaft
+        // images didn't support FeatureLevelRecord, so we can't distinguish 3.2-IV0 from 3.0-IV1.
+        // The least bad way to resolve this is just to pretend that ACLs were in 3.0-IV1.
```
I don't fully understand this comment. Are you saying that Kafka doesn't support clusters that have some servers at software version 3.0.x and some servers, specially the active controller, at software version 3.2.x?

So if the software version of this server is 3.2.x then we need to assume that the software version of every server is 3.2.x?

cmccabe replied:
> I don't fully understand this comment. Are you saying that Kafka doesn't support clusters that have some servers at software version 3.0.x and some servers, specially the active controller, at software version 3.2.x?

no, that is not what this means

> So if the software version of this server is 3.2.x then we need to assume that the software version of every server is 3.2.x?

no

The comment just indicates that we should disallow ACLs in version 3.0, but we won't, because some people might be using them in version 3.2, and we can't distinguish those two versions due to the lack of `metadata.version`.

jsancio replied:
I see. Kafka would only write ACLs records to the snapshot if they existed in the log originally and got replayed to the `AclsImage`. The question is then downgrade. Are you saying that downgrading the `metadata.version` doesn't remove ACLs records.

cmccabe replied:
yes

jsancio commented on metadata/src/main/java/org/apache/kafka/image/AclsImage.java:
```diff
@@ -53,13 +49,15 @@ public Map<Uuid, StandardAcl> acls() {
         for (Entry<Uuid, StandardAcl> entry : acls.entrySet()) {
             StandardAclWithId aclWithId = new StandardAclWithId(entry.getKey(), entry.getValue());
-            batch.add(new ApiMessageAndVersion(aclWithId.toRecord(), (short) 0));
+            writer.write(0, aclWithId.toRecord());
```
This comment applies to a lot of places.

Should we change `0` to a symbol? This would make it a little easier to identify all of the places that need to change if a developers added a new metadata record version.

cmccabe replied:
There are only two places that generate these records. `Image.java` and friends, and the controller. In both cases, we don't have a single version that we care about, but we have to emit the correct record version (which will depend on the current metadata version)

I don't think a constant helps here. In fact I think it would confuse the issue since there are multiple versions we might need to emit.

jsancio commented on metadata/src/main/java/org/apache/kafka/image/FeaturesImage.java:
```diff
@@ -65,24 +64,62 @@ private Optional<Short> finalizedVersion(String feature) {
+        // It is important to write out the metadata.version record first, because it may have an
+        // impact on how we decode records that come after it.
+        //
+        // Note: it's important that this initial FeatureLevelRecord be written with version 0 and
+        // not any later version, so that any modern reader can process it.
```
Interesting. How about writing a similar comment in `FeatureLevelRecord.json` stating that 0 should always be a valid version? Actually, maybe this is true for all metadata records. Zero must always be a valid version for all metadata records.

cmccabe replied:
> Interesting. How about writing a similar comment in FeatureLevelRecord.json stating that 0 should always be a valid version?

There is such a comment in `FeatureLevelRecord.json` already :)

> Actually, maybe this is true for all metadata records. Zero must always be a valid version for all metadata records.

I don't think zero necessarily needs to be valid for all metadata records. We can drop support for old metadata records over time. That would imply that old metadata images would become unreadable, though, so probably not something to do lightly.

Anyway the comment here is not about 0 being supported, but about 0 being the only version we use, at all times.

jsancio commented on metadata/src/main/java/org/apache/kafka/image/ImageWriter.java:
```diff
@@ -0,0 +1,56 @@
+     *                      For convenience, this is an int.
+     * @param message       The message of the record to write out.
+     */
+    default void write(int version, ApiMessage message) {
```
Why is this convenient? Is this because we use the literal `0` which has a type of `int`? If we replace the literal zero with a symbol that has a type of `short`, do we still need this convenience?

cmccabe replied:
It's convenient because writing `(short) 0` and so on is annoying, and Java's typechecker can't recognize that 0 fits in a short apparently (or at least not the version of Java we are using currently)

> If we replace the literal zero with a symbol that has a type of short, do we still need this convenience?

we are probably not going to be using constants here (see above comment)

jsancio commented on metadata/src/main/java/org/apache/kafka/image/ImageWriterOptions.java:
```diff
@@ -0,0 +1,80 @@
+
+        public Builder setMetadataVersion(MetadataVersion metadataVersion) {
+            if (metadataVersion.isLessThan(MetadataVersion.MINIMUM_BOOTSTRAP_VERSION)) {
+                // When writing an image, all versions less than 3.3-IV0 are treated as 3.0-IV1.
+                // This is because those versions don't support FeatureLevelRecord.
+                setRawMetadataVersion(MetadataVersion.MINIMUM_KRAFT_VERSION);
```
I have a couple of questions about this.

First, it looks like the controller only writes the metadata version to the log if it is greater than the minimum bootstrap version so this should never be true, right?
```
              // Write the metadata.version first                                                                           
              if (!wroteVersion) {                                                                                          
                  if (metadataVersion.isAtLeast(minimumBootstrapVersion)) {                                                 
                      wroteVersion = true;                                                                                  
                      return Collections.singletonList(new ApiMessageAndVersion(new FeatureLevelRecord()                    
                              .setName(MetadataVersion.FEATURE_NAME)                                                        
                              .setFeatureLevel(metadataVersion.featureLevel()), FEATURE_LEVEL_RECORD.lowestSupportedVersion(  )));                                                                                                                      
                  }                                                                                                         
              }
```

Second, does this mean that the user cannot incrementally enable features introduced between version 3.0 and 3.3. They have to set the version to 3.3. It doesn't look like the controller checks this when accepting metadata version updates. In other words it looks like the use can set the version 3.2 but it won't have an effect on the metadata version of the cluster. Is this accurate?

cmccabe replied:
> First, it looks like the controller only writes the metadata version to the log if it is greater than the minimum bootstrap version 

greater *or equal*

> so this should never be true, right?

It will never be true if the metadata version is less than 3.3-IV0. Otherwise it will become true when we write that record

> Second, does this mean that the user cannot incrementally enable features introduced between version 3.0 and 3.3. They have to set the version to 3.3.

Yes. We goofed. `FeatureLevelRecord` should have been introduced much earlier than it was. But we can't fix the past.

> It doesn't look like the controller checks this when accepting metadata version updates. In other words it looks like the use can set the version 3.2 but it won't have an effect on the metadata version of the cluster. Is this accurate?

No. The controller does not allow setting any metadata.version that is less than 3.3-IV0

Those old versions are supported for the purpose of rolling upgrade only. You can't create a new cluster with them (through normal means) and you can't downgrade to them.

jsancio commented on metadata/src/main/java/org/apache/kafka/image/ImageWriterOptions.java:
```diff
@@ -0,0 +1,80 @@
+    public void handleLoss(String loss) {
+        lossHandler.accept(new UnwritableMetadataException(metadataVersion, loss));
+    }
+}
```
Minor; add new line character.

cmccabe replied:
ok

jsancio reviewed (commented):
Partial review. Thanks for the changes @cmccabe

jsancio commented on metadata/src/main/java/org/apache/kafka/image/FeaturesImage.java:
```diff
@@ -65,24 +64,43 @@ private Optional<Short> finalizedVersion(String feature) {
         }
+    }
+
+    private void writePreProductionVersion(ImageWriter writer, ImageWriterOptions options) {
```
Instead of `writePreProductionVersion` and `writeProductionVersion` can use a more focused name? How about `handleFeatureLevelNotSupported` and `writeFeatureLevels` respectively?

cmccabe replied:
ok

jsancio commented on metadata/src/main/java/org/apache/kafka/image/ImageProvenance.java:
```diff
@@ -0,0 +1,100 @@
+    private final long highestTimestamp;
+
+    public ImageProvenance(
+        String source,
+        long highestOffset,
+        int highestEpoch,
+        long highestTimestamp
```
Can we document all of these fields? I specifically don't understand how to use the `source` field.

cmccabe replied:
I removed this from the latest

jsancio commented on core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala:
```diff
@@ -179,8 +184,13 @@ class BrokerMetadataListener(
+      } else {
+        val newProvenance = new ImageProvenance("generated snapshot",
+          _highestOffset, _highestEpoch, _highestTimestamp)
+        val newImage = _delta.apply(newProvenance)
```
Minor but how about `snapshotImage` instead of `newImage`.

jsancio commented on metadata/src/main/java/org/apache/kafka/image/writer/ImageReWriter.java:
```diff
@@ -0,0 +1,67 @@
+    private boolean closed = false;
+    private MetadataImage image = null;
+
+    public ImageReWriter(
+        String newSource,
+        MetadataDelta delta
+    ) {
```
Hmm. Maybe I missed it but doesn't look like this class is used anywhere.

I was interested in its usage because it wasn't clear to me when it would be useful to pass in the delta. Instead of starting with a empty delta from an empty or non-empty image.

jsancio commented on metadata/src/main/java/org/apache/kafka/image/writer/ImageWriter.java:
```diff
@@ -0,0 +1,65 @@
+/**
+ * Writes out a metadata image.
+ */
+public interface ImageWriter {
```
It looks like this class can extend from `AutoCloseable`.

jsancio commented on metadata/src/test/java/org/apache/kafka/metadata/MockRandom.java:
```diff
@@ -15,16 +15,26 @@
  */
 
-package org.apache.kafka.metadata;
+package org.apache.kafka.server.util;
```
This package doesn't match the file path. Did you mean to move this file to a different project and package?

jsancio commented on core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala:
```diff
@@ -390,14 +404,18 @@ class BrokerMetadataListener(
+      val records = new util.ArrayList[ApiMessageAndVersion]
+      val writer = new RecordListWriter(records)
+      val options = new ImageWriterOptions.Builder().
+        setMetadataVersion(_image.features().metadataVersion()).
```
This comment also applies to it used in the snapshotter.

Outside of tests it looks like this option is always set to metadata version in the metadata image. What do you think about changing the API so that setting the metadata version is optional. If the metadata version is not set then `MetadataImage.write` will set the metadata version to the version contained in the image?

cmccabe replied:
let's defer this to later. there could be some complexity in doing the fallback.

jsancio reviewed (commented):
Thanks for the changes @cmccabe . Left a few comments.

cmccabe committed:
MINOR: Introduce ImageWriter and ImageWriterOptions

This PR adds a new ImageWriter interface which replaces the generic Consumer interface which
accepted lists of records. It is better to do batching in the ImageWriter than to try to deal with
that complexity in the MetadataImage#write functions, especially since batching is not semantically
meaningful in KRaft snapshots. The new ImageWriter interface also supports freeze and close, which
more closely matches the semantics of the underlying Raft classes.

The PR also adds an ImageWriterOptions class which we can use to pass parameters to control how the
new image is written. Right now, the parameters that we are interested in are the target metadata
version (which may be more or less than the original image's version) and a handler function which
is invoked whenever metadata is lost due to the target version.

Convert over the MetadataImage#write function (and associated functions) to use the new ImageWriter
and ImageWriterOptions. In particular, we now have a way to handle metadata losses by invoking
ImageWriterOptions#handleLoss. This allows us to handle writing an image at a lower version, for
the first time. This support is still not enabled externally by this PR, though. That will come in
a future PR.

Get rid of the use of SOME_RECORD_TYPE.highestSupportedVersion() in several places. In general, we
do not want to "silently" change the version of a record that we output, just because a new version
was added. We should be explicit about what record version numbers we are outputting.

Implement ProducerIdsDelta#toString, to make debug logs look better.

Move MockRandom to the server-common package so that other internal broker packages can use it.

jsancio reviewed (approved):
LGTM. All of the test failures seem unrelated.
